Check for consistent manifold ids in 2D when creating a Triangulation

masterleinad opened the pull request:
Fixes #6397, i.e. replaces `invalid_manifold_id` by ` flat_manifold_id`. I decided to deprecate the constant rather than removing it outright. Even if it is very unlikely that anyone has used this constant outside the internals of the library, I prefer to make sure that we stay backward-compatible here as well.

Additionally, this PR introduces another check for consistent manifold ids when creating a two-dimensional `Triangulation`. The corresponding check for the three-dimensional case is already in place.

drwells commented on source/grid/tria.cc:
```diff
@@ -1257,7 +1257,18 @@ namespace internal
+      int,
+      << "In SubCellData the line info of the line with vertex indices " << arg1
+      << " and " << arg2
+      << " appears mutiply with different (valid) manifold ids. "
```
should be 'multiply' or 'multiple times'. I think 'multiple times' is easier to understand.

masterleinad replied:
Well, if a native speaker is confused by "multiply", I am happy to rephrase.

drwells commented:
We can use the same line exceptions in the 3D version of the function too. Would you mind doing that? If not we can just open an issue for it and fix it later.

drwells commented on source/grid/tria.cc:
```diff
@@ -2192,15 +2227,10 @@ namespace internal
-                    else
-                      {
-                        line->set_manifold_id(subcell_line.manifold_id);
-                      }
```
Now that this `else` statement has been redone the previous block doesn't really make sense: whether or not a manifold id is set has nothing to do with the boundary id of the line. Would you mind refactoring this check (i.e., lines 2202 through 2334)?

drwells commented:
This patch is fortuitously timed: I ended up adding an option to copy manifold ids to `GridGenerator::extrude_triangulation` as part of #6863 and the remaining difficulty is setting manifold ids on lines: the new checks should help!

masterleinad commented:
Simplifying the check required that we set `numbers::internal_face_boundary_id` for the faces and lines we want to set a manifold id for in `GridGenerator::merge_triangulations`. This makes sense to me.

drwells commented on source/grid/tria.cc:
```diff
@@ -2196,41 +2196,14 @@ namespace internal
+            // Assert that only exterior lines are given a boundary indicator
+            AssertThrow(
+              subcell_line.boundary_id == numbers::internal_face_boundary_id ||
+                line->boundary_id() != numbers::internal_face_boundary_id,
```
Thanks for cleaning this up.

I might be missing something, but I am not convinced that this assertion is correct. I think it is possible for it to pass when `subcell_line.boundary_id == numbers::internal_face_boundary_id` and `line->boundary_id() != numbers::internal_face_boundary_id`. If so then we just assigned a boundary line to have an interior id. Is this right?

masterleinad replied:
Yes, you're right. It should read
```
if (subcell_line.boundary_id != numbers::internal_face_boundary_id)
  {
    AssertThrow(
      line->boundary_id() != numbers::internal_face_boundary_id,
      ExcInteriorLineCantBeBoundary(line->vertex_index(0),
                                    line->vertex_index(1),
                                    subcell_line.boundary_id));
    line->set_boundary_id_internal(subcell_line.boundary_id);
  }
```
I fixed that now.

drwells commented:
The test failure is interesting: the input file sets (what are ultimately treated as) boundary ids for all lines. I believe that we previously ignored these values, so we will have to modify the UCD input parser to ignore them again.

masterleinad commented:
That's annoying. Unfortunately, I don't see a simple way to modify `GridIn::read_ucd` since we don't know if a cell is at the boundary or in the interior.
I would simply ignore boundary ids for internal faces and leave a note in `GridGenerator::create_triangulation` and `GridIn::read_ucd`.

masterleinad commented:
I restored the old behavior where we allow to set boundary ids on internal faces, but not `numbers::internal_face_boundary_id` for boundary faces. For this PR, this seems to be the most appropriate way. I opened #7033 for a better solution to this.

masterleinad commented:
This should finally work. Fixing the `read_ucd_01` test was easier than expected. We were just missing to set `boundary_id` to `internal_face_boundary_id` when the `material id` in the input file should be used as `manifold id`. Still, we might want to address #7033 in a different PR.

masterleinad commented:
For the `codim` tests, I had to set `internal_face_boundary_id` for the internal faces.

drwells commented on source/grid/grid_generator.cc:
```diff
@@ -5196,6 +5210,8 @@ namespace GridGenerator
                               map_vert_index[face->line(e)->vertex_index(0)])))
                           {
+                            subcell_data.boundary_lines[i].boundary_id =
+                              numbers::internal_face_boundary_id;
```
This is quite clever!

drwells commented on source/grid/grid_in.cc:
```diff
@@ -844,11 +852,19 @@ GridIn<dim, spacedim>::read_ucd(std::istream &in,
                                numbers::internal_face_boundary_id));
 
           if (apply_all_indicators_to_manifolds)
```
I read through the list of changes again and this is the only part I don't understand. Would you add a comment to help explain why this works?

drwells commented:
Nice work. This was tricky.

drwells commented:
I thought of one more thing we should do before we merge: we are now much stricter in not setting boundary ids on interior edges. Would you mind adding a changelog entry describing this?

masterleinad commented:
> Would you mind adding a changelog entry describing this?

No, I added one and updated a comment in `GridGenerator::extract_boundary_mesh`. This PR grew much larger than expected...

drwells commented:
These kinds of geometry/topology problems are always hard.

I suspect this might break some more stuff, but only if it was already silently broken (e.g., by setting internal boundary ids), so lets go ahead and merge.

bangerth commented:
Good job deprecating rather than removing the constant, btw. ASPECT turns out to use these constants.
